esp-wifi/src/preempt.rs

```diff
@@ -0,0 +1,147 @@
+/// This module allows hooking `esp-wifi` into an external scheduler, instead
+/// of using the integrated one as provided by the `preempt` module.
+///
+/// In order to hook up an external scheduler, enable the `preempt-extern`
+/// feature, implement the `Scheduler` trait for a struct, and create the
+/// necessary `extern` functions using the `scheduler_impl!()` macro.
+///
+/// Example:
+///
+/// ```ignore
+/// use esp_wifi::preempt::Scheduler;
+///
+/// struct MyScheduler {}
+///
+/// impl Scheduler for MyScheduler {
+///     // impl goes here
+/// }}
+///
+/// esp_wifi::scheduler_impl!(static SCHEDULER: MyScheduler = MyScheduler {});
+/// ```
+use core::ffi::c_void;
+
+pub trait Scheduler: Send + Sync + 'static {
+    /// This function is called by `esp-wifi` when starting up the WiFi stack.
+    /// The passed `timer` can be used for e.g., implementing a periodic timer.
+    fn setup(&self, timer: crate::TimeBase);
```

Comment:
Would be nice to remove this from the trait some how, but I don't think we can because the scheduler has to live for `'static` which means we can't at runtime construct it with the timer that it needs :thinking:

Reply:
I feel the same. Maybe we could just keep it out, and document that a certain tick should be set up? For the builtin scheduler, this could be set up manually.

I'd assume that usually, a system would first set up the scheduler et al, then start setting up the Network/WiFi. At the point in system initialization where some `esp_wifi::init()` is called, the scheduler is probably already running and ticking happily. I'd think that this `setup()` is more for completeness, and for this builtin scheduler, and maybe to set up `esp_wifi` specific settings in an already running scheduler.

(I actually don't think the time-slicing is needed, at least in Ariel I'm not setting that up. WiFi still works as before. But I wouldn't want to lean too far out of the window here.)

Reply:
> I feel the same. Maybe we could just keep it out, and document that a certain tick should be set up? For the builtin scheduler, this could be set up manually.

This sounds reasonable to me. You still need something to signal that the scheduler should start (since we have a matching `disable`), but the timer should probably be an implementation detail.

We can probably make esp-wifi take the scheduler as an init argument, that way we don't have to eat a timer that would be left unused in Ariel OS.

Reply:
ok, that would probably require splitting the init function:

```
// old
esp_wifi::init(timer, rng, peripherals.RADIO_CLK) {
  init_builtin_sched(timer);
  init_rest(rng, ...);
}
// new
esp_wifi::init_builtin_sched(timer) {
// call scheduler `setup()`
// set-up `timer` tick but disabled, `Scheduler::setup()` enables the isr
}
esp_wifi::init_rest(rng, peripherals.RADIO_CLK) {
// as now but without timer
}
```

or, split the function in the public API:

```
// first
esp_wifi::init_builtin_sched(timer);
// then
esp_wifi::init(rng, peripherals.RADIO_CLK);
```

Preferences? And help with names please ;)

Reply:
> We can probably make esp-wifi take the scheduler as an init argument, that way we don't have to eat a timer that would be left unused in Ariel OS. Although this is a bit of a stretch and changes the concept a bit, so maybe not.

Let's tackle this bit later, once we've pulled our scheduler out.

Whilst the scheduler here is still built in, we can just call the setup function somewhere in `esp_wifi::init` as the timer is there already (if you don't mind added a comment about it too, that would be great!). Then replace `setup` with `enable` as @bugadani suggested and I think we should be good?

Reply:
> Whilst the scheduler here is still built in, we can just call the setup function somewhere in `esp_wifi::init` as the timer is there already (if you don't mind added a comment about it too, that would be great!).

You mean, call it through the `Scheduler` trait as it is now? Or pull out the timer initialization?

> Then replace `setup` with `enable` as @bugadani suggested and I think we should be good?

Where did he suggest that (I don't find it)? Rename the trait's `setup()` to `enable()`?

Thanks for clarifying, I think it has been a long day here. :)

Reply:
Sorry I was referencing @bugadani with

> You still need something to signal that the scheduler should start (since we have a matching disable), but the timer should probably be an implementation detail.

> You mean, call it through the Scheduler trait as it is now? Or pull out the timer initialization?

Pull out the timer initialization from the scheduler trait, and dump it in `esp_wifi::init` for now - we can find a proper home for it later.